File: Week11/Leetcode-78-Subsets.py
# dfs(using recursion)

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(List,index) :
            if len(List) == len(nums) :
                return
            
            for i in range(index, len(nums)) :
                List.append(nums[i])
                result.append(List[:])
                dfs(List,i+1)
                List.pop()
            
            return

        result = [[]]
        dfs([],0)

        
        return result

File: graph/Leetcode-78-Subsets/Leetcode-78-Subsets.cpp
class Solution {
public:
    vector<vector<int>> result;
    
    vector<vector<int>> subsets(vector<int>& nums) {
       vector<int> tmp;
        appendResult(nums,tmp,0);
        return result;
    }
    
    void appendResult(vector<int>&nums,vector<int>& tmp,int start){
        if(start==nums.size()){
            result.push_back(tmp);
            return;
        }
        tmp.push_back(nums[start]);
        appendResult(nums,tmp,start+1);
        tmp.pop_back();
        appendResult(nums,tmp,start+1);
    }
};